src/serde/read_cache_lookup.rs

```diff
@@ -0,0 +1,232 @@
+        *self.count.entry(right.0.clone()).or_insert(0) += 1;
+
+        let new_root_hash = hash_blobs(&[&[2], &left.0, &right.0]);
+
+        self.parent_lookup
+            .entry(left.0)
+            .or_default()
+            .push((new_root_hash.clone(), 0));
+
+        self.parent_lookup
+            .entry(right.0)
+            .or_default()
+            .push((new_root_hash.clone(), 1));
+
+        self.push(new_root_hash);
+    }
+
+    /// return the list of minimal-length paths to the given hash which will serialize to no larger
+    /// than the given size (or an empty list if no such path exists)
+    pub fn find_paths(&self, id: &Bytes32, serialized_length: usize) -> Vec<Vec<u8>> {
+        let mut seen_ids = HashSet::<&Bytes32>::default();
+        let mut possible_responses = vec![];
+        if serialized_length < 3 {
+            return possible_responses;
+        }
+        let max_bytes_for_path_encoding = serialized_length - 2; // 1 byte for 0xfe, 1 min byte for savings
+        let max_path_length = max_bytes_for_path_encoding * 8 - 1;
+        seen_ids.insert(id);
+        let mut partial_paths = vec![(id.clone(), vec![])];
+
+        loop {
+            if partial_paths.is_empty() {
+                break;
+            }
+            let mut new_partial_paths = vec![];
+            for (node, path) in partial_paths.iter_mut() {
+                if *node == self.root_hash {
+                    possible_responses.push(reversed_path_to_vec_u8(path));
+                    continue;
+                }
+
+                let parents = self.parent_lookup.get(node);
+                if let Some(items) = parents {
+                    for (parent, direction) in items.iter() {
+                        if *(self.count.get(parent).unwrap_or(&0)) > 0 && !seen_ids.contains(parent)
+                        {
+                            let mut new_path = path.clone();
+                            new_path.push(*direction);
+                            if new_path.len() > max_path_length {
+                                return possible_responses;
+                            }
+                            new_partial_paths.push((parent.clone(), new_path));
+                        }
+                        seen_ids.insert(parent);
+                    }
+                }
+            }
+            if !possible_responses.is_empty() {
+                break;
+            }
+            partial_paths = new_partial_paths;
+        }
+        possible_responses
+    }
+
+    /// If multiple paths exist, the lexigraphically smallest one will be returned.
+    pub fn find_path(&self, id: &Bytes32, serialized_length: usize) -> Option<Vec<u8>> {
+        let mut paths = self.find_paths(id, serialized_length);
+        if !paths.is_empty() {
+            paths.sort();
+            paths.truncate(1);
+            paths.pop()
+        } else {
+            None
+        }
+    }
+}
+
+fn reversed_path_to_vec_u8(path: &[u8]) -> Vec<u8> {
```

Comment:
could you add a comment explaining what this function does?

Reply:
Done.